### PR TITLE
Partition: add key-partiton document

### PR DESCRIPTION
**Pull request by TonsnakeLin**
### First-time contributors' checklist 

- [ ] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### What is changed, added or deleted? (Required)
basic-features.md, mysql-compatibility.md, partitioned-table.md

### Which TiDB version(s) do your changes apply to? (Required)
7.0

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v7.0 (TiDB 7.0 versions)
- [ ] v6.6 (TiDB 6.6 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.4 (TiDB 6.4 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?
https://github.com/pingcap/tidb/pull/41365

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [x] Change aliases
- [ ] Need modification after applied to another branch 
- [x] Might cause conflicts after applied to another branch

### CODE REVIEW

File: partitioned-table.md [hunk 1]
```diff
@@ -817,6 +908,22 @@ alter table members optimize partition p0;
...
+### Key 分区管理
+
+目前 TiDB 中 Key 分区支持的分区管理语句只有有 `ALTER TABLE ... TRUNCATE PARTITION`。
+
+对于暂不支持的分区管理语句，TiDB 会返回错误。
+
+{{< copyable "sql" >}}
+
+```sql
+alter table members optimize partition p0;
```
Comment (qiancai) [inline partitioned-table.md:920]:
```suggestion
ALTER TABLE members OPTIMIZE PARTITION p0;
```

File: partitioned-table.md [hunk 2]
```diff
@@ -817,6 +908,22 @@ alter table members optimize partition p0;
 ERROR 8200 (HY000): Unsupported optimize partition
 ```
```
[the remark this answers is not in the document]
Comment (qiancai) [reply partitioned-table.md:918]: @TonsnakeLin PTAL 是否需要添加

[the remark this answers is not in the document]
Comment (TonsnakeLin) [reply partitioned-table.md:918]: I'll add it

File: partitioned-table.md [hunk 3]
```diff
@@ -545,6 +548,92 @@ MOD(YEAR('2005-09-01'),4)
 =  1
 ```
 
+### Key 分区
+
+TiDB 从 v7.0.0 开始支持 Key 分区。在 v7.0.0 之前的版本中，创建 Key 分区表时，TiDB 会将其创建为非分区表并给出告警。 
+
+Key 分区与 Hash 分区都可以保证将数据均匀地分散到一定数量的分区里面，区别是 Hash 分区只能根据一个指定的整数表达式或字段进行分区，而 Key 分区可以根据字段列表进行分区，且 Key 分区的分区字段不局限于整数类型。TiDB Key 分区表的 Hash 算法与 MySQL 不一样，因此表的数据分布也不一样。
+
+使用 Key 分区时，你需要在 `CREATE TABLE` 后面添加 `PARTITION BY HASH (columList)`，其中 `columnList` 是字段列表，可以包含一个或多个字段。每个字段的类型可以是除 `BLOB`、`JSON`、`GEOMETRY` 之外的任意类型。此外，你很可能还需要加上 `PARTITIONS num`，其中 `num` 是一个正整数，表示将表划分多少个分区。
```
Comment (bb7133) [inline partitioned-table.md]:
```suggestion
使用 Key 分区时，你需要在 `CREATE TABLE` 后面添加 `PARTITION BY HASH (columList)`，其中 `columnList` 是字段列表，可以包含一个或多个字段。每个字段的类型可以是除 `BLOB`、`JSON`、`GEOMETRY` 之外的任意类型（请注意 TiDB 不支持 `GEOMETRY` 类型）。此外，你很可能还需要加上 `PARTITIONS num`，其中 `num` 是一个正整数，表示将表划分多少个分区。
```

**Review by bb7133 — APPROVED** [review]
LGTM

File: partitioned-table.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (qiancai) [reply partitioned-table.md:918]: 对于 `HASH` 分区表，目前也是只支持 `ALTER TABLE ... TRUNCATE PARTITION` 分区管理语句吗？

File: mysql-compatibility.md [hunk 4]
```diff
@@ -112,10 +112,8 @@ TiDB 中，所有支持的 DDL 变更操作都是在线执行的。与 MySQL 相
 * TiDB 中，`ALGORITHM={INSTANT,INPLACE,COPY}` 语法只作为一种指定，并不更改 `ALTER` 算法，详情参阅 [`ALTER TABLE`](/sql-statements/sql-statement-alter-table.md)。
 * 不支持添加或删除 `CLUSTERED` 类型的主键。要了解关于 `CLUSTERED` 主键的详细信息，请参考[聚簇索引](/clustered-indexes.md)。
 * 不支持指定不同类型的索引 (`HASH|BTREE|RTREE|FULLTEXT`)。若指定了不同类型的索引，TiDB 会解析并忽略这些索引。
-* 分区表支持 `HASH`、`RANGE` 和 `LIST` 分区类型。对于不支持的分区类型，TiDB 可能会报 `Warning: Unsupported partition type %s, treat as normal table` 错误，其中 `%s` 为不支持的具体分区类型。
+* 分区表支持 `HASH`、`RANGE` 、`LIST` 和 `KEY` 分区类型。`KEY` 分区类型暂不支持分区字段列表为空的语句。对于不支持的分区类型，TiDB 可能会报 `Warning: Unsupported partition type %s, treat as normal table` 错误，其中 `%s` 为不支持的具体分区类型。
```
Comment (hfxsd) [inline mysql-compatibility.md:115]:
```suggestion
* 分区表支持 `HASH`、`RANGE`、`LIST` 和 `KEY` 分区类型。`KEY` 分区类型暂不支持分区字段列表为空的语句。对于不支持的分区类型，TiDB 可能会报 `Warning: Unsupported partition type %s, treat as normal table` 错误，其中 `%s` 为不支持的具体分区类型。
```

File: partitioned-table.md [hunk 5]
```diff
@@ -545,6 +548,92 @@ MOD(YEAR('2005-09-01'),4)
...
+    store_id INT
+)
+
+PARTITION BY KEY(fname, store_id)
+PARTITIONS 4;
+```
+
+目前，TiDB 不支持分区字段列表 `PARTITION BY KEY` 为空的 Key 分区表。下面的语句将创建一个非分区表，并向客户端返回 `Unsupported partition type RANGE, treat as normal table` 警告。
+
+{{< copyable "sql" >}}
```
Comment (hfxsd) [inline partitioned-table.md:620]:
```suggestion
```

File: partitioned-table.md [hunk 6]
```diff
@@ -545,6 +548,92 @@ MOD(YEAR('2005-09-01'),4)
...
+    store_id INT
+)
+
+PARTITION BY KEY(fname)
+PARTITIONS 4;
+```
+
+你还可以根据多列字段创建 Key 分区表。下面的语句按 `fname`、`store_id` 将表分成 4 个分区：
+
+{{< copyable "sql" >}}
```
Comment (hfxsd) [inline partitioned-table.md:601]:
```suggestion
```

File: partitioned-table.md [hunk 7]
```diff
@@ -545,6 +548,92 @@ MOD(YEAR('2005-09-01'),4)
...
+
+PARTITION BY KEY(store_id)
+PARTITIONS 4;
+```
+
+如果不指定 `PARTITIONS num`，默认的分区数量为 1。
+
+你也可以根据 VARCHAR 等非整数字段创建 Key 分区表。下面的语句按 `fname` 将表分成 4 个分区：
+
+{{< copyable "sql" >}}
```
Comment (hfxsd) [inline partitioned-table.md:582]:
```suggestion
```

File: partitioned-table.md [hunk 8]
```diff
@@ -545,6 +548,92 @@ MOD(YEAR('2005-09-01'),4)
...
+
+TiDB 从 v7.0.0 开始支持 Key 分区。在 v7.0.0 之前的版本中，创建 Key 分区表时，TiDB 会将其创建为非分区表并给出告警。 
+
+Key 分区与 Hash 分区都可以保证将数据均匀地分散到一定数量的分区里面，区别是 Hash 分区只能根据一个指定的整数表达式或字段进行分区，而 Key 分区可以根据字段列表进行分区，且 Key 分区的分区字段不局限于整数类型。TiDB Key 分区表的 Hash 算法与 MySQL 不一样，因此表的数据分布也不一样。
+
+使用 Key 分区时，你需要在 `CREATE TABLE` 后面添加 `PARTITION BY KEY (columList)`，其中 `columnList` 是字段列表，可以包含一个或多个字段。每个字段的类型可以是除 `BLOB`、`JSON`、`GEOMETRY` 之外的任意类型（请注意 TiDB 不支持 `GEOMETRY` 类型）。此外，你很可能还需要加上 `PARTITIONS num`，其中 `num` 是一个正整数，表示将表划分多少个分区。
+
+下面的语句将创建一个 Key 分区表，按 `store_id` 分成 4 个分区：
+
+{{< copyable "sql" >}}
```
Comment (hfxsd) [inline partitioned-table.md:561]:
```suggestion
```

File: partitioned-table.md [hunk 9]
```diff
@@ -817,6 +916,22 @@ alter table members optimize partition p0;
 ERROR 8200 (HY000): Unsupported optimize partition
 ```
 
+### Key 分区管理
+
+目前 TiDB 中 Key 分区支持的分区管理语句只有 `ALTER TABLE ... TRUNCATE PARTITION`。
```
Comment (hfxsd) [inline partitioned-table.md:921]: 加个链接到 `ALTER TABLE ... TRUNCATE PARTITION`？

File: partitioned-table.md [hunk 10]
```diff
@@ -817,6 +916,22 @@ alter table members optimize partition p0;
...
+
+对于暂不支持的分区管理语句，TiDB 会返回错误。
+
+{{< copyable "sql" >}}
+
```
Comment (hfxsd) [inline partitioned-table.md:926]:
```suggestion
```

**Review by hfxsd — APPROVED** [review]
Rest LGTM

File: partitioned-table.md [hunk 11]
```diff
@@ -817,6 +908,20 @@ alter table members optimize partition p0;
 ERROR 8200 (HY000): Unsupported optimize partition
 ```
 
+### Key 分区管理
+
+目前 TiDB 中 Key 分区支持的分区管理语句只有 `ALTER TABLE ... TRUNCATE PARTITION`。
```
Comment (qiancai) [inline partitioned-table.md:913]:
```suggestion
目前 TiDB 中 Key 分区支持的分区管理语句只有 `ALTER TABLE ... TRUNCATE PARTITION`。

```sql
ALTER TABLE members TRUNCATE PARTITION p0;
```

```
Query OK, 0 rows affected (0.03 sec)
```

```

File: partitioned-table.md (hunk 9 above)
Comment (qiancai) [reply partitioned-table.md:913]: 没有单独对应的文档或章节，要不给一个实例吧

File: partitioned-table.md [hunk 12]
```diff
@@ -703,13 +790,17 @@ Empty set (0.00 sec)
...
+在 Key 分区中 NULL 值的处理与 Hash 分区一致：如果分区字段的值为 NULL，它会被当作 0 值处理。
+
 ## 分区管理
 
 对于 `LIST` 和 `RANGE` 分区表，通过 `ALTER TABLE <表名> ADD PARTITION (<分区说明>)` 或 `ALTER TABLE <表名> DROP PARTITION <分区列表>` 语句，可以执行添加和删除分区的操作。
 
 对于 `LIST` 和 `RANGE` 分区表，暂不支持 `REORGANIZE PARTITION` 语句。
 
-对于 `HASH` 分区表，暂不支持 `COALESCE PARTITION` 和 `ADD PARTITION` 语句。
+对于 `HASH` 和 `KEY` 分区表，暂不支持 `COALESCE PARTITION` 和 `ADD PARTITION` 语句。
```
Comment (qiancai) [inline partitioned-table.md:803]:
```suggestion
对于 `HASH` 和 `KEY` 分区表，目前只支持 `ALTER TABLE ... TRUNCATE PARTITION` 分区语句，不支持 `COALESCE PARTITION` 和 `ADD PARTITION` 语句。
```

File: partitioned-table.md [hunk 13]
```diff
@@ -1338,9 +1451,9 @@ YEARWEEK()
 
 ### 兼容性
 
-目前 TiDB 支持 Range 分区、List 分区、List COLUMNS 分区和 Hash 分区，其它的 MySQL 分区类型（例如 Key 分区）尚不支持。
+目前 TiDB 支持 Range 分区、List 分区、List COLUMNS 分区、Hash 分区和 Key 分区，其它的 MySQL 分区类型尚不支持。
```
Comment (qiancai) [inline partitioned-table.md:1454]:
```suggestion
目前 TiDB 支持 Range 分区、Range Columns 分区、List 分区、List COLUMNS 分区、Hash 分区和 Key 分区，其它的 MySQL 分区类型尚不支持。
```

**Review by qiancai — APPROVED** [review]

File: partitioned-table.md [hunk 14]
```diff
@@ -555,6 +636,12 @@ MOD(YEAR('2005-09-01'),4)
 
 - 对于 MySQL Linear Hash 分区的其他 SQL 语句，TiDB 将正常返回对应的 Hash 分区的查询结果。但当分区数不是 2 的幂（意味着分区表中行的分布情况与 MySQL 不同）时，[分区选择](#分区选择)、`TRUNCATE PARTITION`、`EXCHANGE PARTITION` 返回的结果将和 MySQL 有所差异。
 
+### TiDB 对 Linear Key 分区的处理
+
+TiDB 从 v7.0.0 开始支持 Key 分区，并支持解析 MySQL 的 `PARTITION BY LINEAR HASH` 语法，但会忽略其中的 `LINEAR` 关键字，只采用非线性 Hash 算法。
```
Comment (qiancai) [inline partitioned-table.md:641]:
```suggestion
TiDB 从 v7.0.0 开始支持 Key 分区，并支持解析 MySQL 的 `PARTITION BY LINEAR  KEY` 语法，但会忽略其中的 `LINEAR` 关键字，只采用非线性 Hash 算法。
```

File: partitioned-table.md [hunk 15]
```diff
@@ -545,6 +548,84 @@ MOD(YEAR('2005-09-01'),4)
 =  1
 ```
 
+### Key 分区
+
+TiDB 从 v7.0.0 开始支持 Key 分区。在 v7.0.0 之前的版本中，创建 Key 分区表时，TiDB 会将其创建为非分区表并给出告警。 
+
+Key 分区与 Hash 分区都可以保证将数据均匀地分散到一定数量的分区里面，区别是 Hash 分区只能根据一个指定的整数表达式或字段进行分区，而 Key 分区可以根据字段列表进行分区，且 Key 分区的分区字段不局限于整数类型。TiDB Key 分区表的 Hash 算法与 MySQL 不一样，因此表的数据分布也不一样。
+
+使用 Key 分区时，你需要在 `CREATE TABLE` 后面添加 `PARTITION BY KEY (columList)`，其中 `columnList` 是字段列表，可以包含一个或多个字段。每个字段的类型可以是除 `BLOB`、`JSON`、`GEOMETRY` 之外的任意类型（请注意 TiDB 不支持 `GEOMETRY` 类型）。此外，你很可能还需要加上 `PARTITIONS num`，其中 `num` 是一个正整数，表示将表划分多少个分区；或者加上指定名称的分区定义，例如 `(PARTITION pOne, PARTITION pTwo)`。
```
Comment (qiancai) [inline partitioned-table.md:557]:
```suggestion
使用 Key 分区时，你需要在 `CREATE TABLE` 后面添加 `PARTITION BY KEY (columList)`，其中 `columnList` 是字段列表，可以包含一个或多个字段。每个字段的类型可以是除 `BLOB`、`JSON`、`GEOMETRY` 之外的任意类型（请注意 TiDB 不支持 `GEOMETRY` 类型）。此外，你很可能还需要加上 `PARTITIONS num`，其中 `num` 是一个正整数，表示将表划分多少个分区；或者加上分区名的定义，例如，加上 `(PARTITION p0, PARTITION p1)` 代表将表划分为两个分区，分区名为 p0 和 p1。
```

File: partitioned-table.md [hunk 16]
```diff
@@ -545,6 +548,84 @@ MOD(YEAR('2005-09-01'),4)
 =  1
 ```
 
+### Key 分区
+
+TiDB 从 v7.0.0 开始支持 Key 分区。在 v7.0.0 之前的版本中，创建 Key 分区表时，TiDB 会将其创建为非分区表并给出告警。 
+
+Key 分区与 Hash 分区都可以保证将数据均匀地分散到一定数量的分区里面，区别是 Hash 分区只能根据一个指定的整数表达式或字段进行分区，而 Key 分区可以根据字段列表进行分区，且 Key 分区的分区字段不局限于整数类型。TiDB Key 分区表的 Hash 算法与 MySQL 不一样，因此表的数据分布也不一样。
+
+使用 Key 分区时，你需要在 `CREATE TABLE` 后面添加 `PARTITION BY KEY (columList)`，其中 `columnList` 是字段列表，可以包含一个或多个字段。每个字段的类型可以是除 `BLOB`、`JSON`、`GEOMETRY` 之外的任意类型（请注意 TiDB 不支持 `GEOMETRY` 类型）。此外，你很可能还需要加上 `PARTITIONS num`，其中 `num` 是一个正整数，表示将表划分多少个分区；或者加上分区名的定义，例如，加上 `(PARTITION p0, PARTITION p1)` 代表将表划分为两个分区，分区名为 p0 和 p1。
```
Comment (qiancai) [inline partitioned-table.md:557]:
```suggestion
创建 Key 分区表时，你需要在 `CREATE TABLE` 后面添加 `PARTITION BY KEY (columList)`，其中 `columnList` 是字段列表，可以包含一个或多个字段。每个字段的类型可以是除 `BLOB`、`JSON`、`GEOMETRY` 之外的任意类型（请注意 TiDB 不支持 `GEOMETRY` 类型）。此外，你很可能还需要加上 `PARTITIONS num`，其中 `num` 是一个正整数，表示将表划分多少个分区；或者加上分区名的定义，例如，加上 `(PARTITION p0, PARTITION p1)` 代表将表划分为两个分区，分区名为 p0 和 p1。
```

**Pull request merged**